File: backend/app/schemas/publication.py
"""HTTP shapes for the run-scoped Wikidata Publication seam."""

from __future__ import annotations

from datetime import datetime
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field


class PublicationSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")


class RunPublicationSource(PublicationSchema):
    kind: Literal["run"] = "run"
    projection_source: Literal["legacy", "canonical"]
    approved_only: bool = True


class PreparePublicationRequest(PublicationSchema):
    profile_id: str = Field(min_length=1, max_length=128)
    profile_version: str = Field(min_length=1, max_length=64)
    target: Literal["test", "live"]
    source: RunPublicationSource


class EligibleReleaseSelection(PublicationSchema):
    mode: Literal["eligible_release"]


class EntityKeysSelection(PublicationSchema):
    mode: Literal["entities"]
    entity_keys: list[str] = Field(min_length=1, max_length=500)


ReviewSelection = Annotated[
    EligibleReleaseSelection | EntityKeysSelection,
    Field(discriminator="mode"),
]


class ReviewPublicationCommand(PublicationSchema):
    type: Literal["review"]
    release_id: str = Field(min_length=1, max_length=128)
    expected_release_digest: str = Field(min_length=1, max_length=128)
    selection: ReviewSelection
    decision: Literal["approve", "reject"]
    reason: str = Field(min_length=1, max_length=2_000)


class DryRunPublicationCommand(PublicationSchema):
    type: Literal["dry_run"]
    approval_set_id: str = Field(min_length=1, max_length=128)
    expected_approval_digest: str = Field(min_length=1, max_length=128)


class PublishPublicationCommand(PublicationSchema):
    type: Literal["publish"]
    plan_id: str = Field(min_length=1, max_length=128)
    dry_run_receipt_id: str = Field(min_length=1, max_length=128)
    expected_receipt_digest: str = Field(min_length=1, max_length=128)


class ResumePublicationCommand(PublicationSchema):
    type: Literal["resume"]
    execution_id: str = Field(min_length=1, max_length=128)


class CancelPublicationCommand(PublicationSchema):
    type: Literal["cancel"]
    operation_id: str | None = Field(default=None, max_length=128)
    reason: str | None = Field(default=None, max_length=2_000)


PublicationAdvanceCommand = Annotated[
    ReviewPublicationCommand
    | DryRunPublicationCommand
    | PublishPublicationCommand
    | ResumePublicationCommand
    | CancelPublicationCommand,
    Field(discriminator="type"),
]


class AdvancePublicationRequest(PublicationSchema):
    command: PublicationAdvanceCommand


class PublicationSummaryQuery(PublicationSchema):
    type: Literal["summary"]


class PublicationEntitiesQuery(PublicationSchema):
    type: Literal["entities"]
    release_id: str = Field(min_length=1, max_length=128)
    cursor: str | None = Field(default=None, max_length=2_048)
    limit: int = Field(default=100, ge=1, le=500)
    entity_kind: str | None = Field(default=None, max_length=64)
    review_status: Literal["pending", "approved", "rejected", "stale"] | None = None
    query: str | None = Field(default=None, max_length=256)


class PublicationOperationQuery(PublicationSchema):
    type: Literal["operation"]
    operation_id: str = Field(min_length=1, max_length=128)


class PublicationAuditQuery(PublicationSchema):
    type: Literal["audit"]
    cursor: str | None = Field(default=None, max_length=2_048)
    limit: int = Field(default=100, ge=1, le=500)


PublicationReadQuery = Annotated[
    PublicationSummaryQuery
    | PublicationEntitiesQuery
    | PublicationOperationQuery
    | PublicationAuditQuery,
    Field(discriminator="type"),
]


class ReadPublicationRequest(PublicationSchema):
    query: PublicationReadQuery


class PublicationFindingCounts(PublicationSchema):
    error: int = 0
    warning: int = 0
    info: int = 0


class ReleaseSummary(PublicationSchema):
    release_id: str
    release_digest: str
    revision: int
    created_at: datetime
    entity_count: int
    finding_counts: PublicationFindingCounts


class ApprovalSetSummary(PublicationSchema):
    approval_set_id: str
    approval_digest: str
    release_id: str
    release_digest: str
    status: Literal["pending", "approved", "rejected", "stale"]
    approved_count: int
    rejected_count: int
    pending_count: int
    created_at: datetime


class PlanSummary(PublicationSchema):
    plan_id: str
    plan_digest: str
    release_id: str
    release_digest: str
    approval_set_id: str
    status: Literal["ready", "blocked", "expired", "executed"]
    expires_at: datetime | None
    action_counts: dict[str, int]


class DryRunReceiptSummary(PublicationSchema):
    dry_run_receipt_id: str
    receipt_digest: str
    plan_id: str
    plan_digest: str
    status: Literal["valid", "stale", "failed"]
    checked_at: datetime
    expires_at: datetime | None


class ExecutionSummary(PublicationSchema):
    execution_id: str
    plan_id: str
    status: Literal["queued", "running", "paused", "succeeded", "failed", "cancelled"]
    processed: int
    total: int
    succeeded: int
    failed: int
    skipped: int
    current_entity_label: str | None
    started_at: datetime | None
    finished_at: datetime | None


class PublicationOperation(PublicationSchema):
    operation_id: str
    command: Literal["prepare", "review", "dry_run", "publish", "resume", "cancel"]
    status: Literal["queued", "running", "succeeded", "failed", "cancelled"]
    progress: ExecutionSummary | None
    error: str | None


class PublicationSummary(PublicationSchema):
    publication_id: str
    run_id: str
    profile_id: str
    profile_version: str
    target: Literal["test", "live"]
    status: Literal[
        "preparing",
        "ready_for_review",
        "reviewed",
        "dry_run_ready",
        "publishing",
        "paused",
        "completed",
        "cancelled",
        "failed",
    ]
    source_current: bool
    current_release: ReleaseSummary
    approval_set: ApprovalSetSummary | None
    plan: PlanSummary | None
    dry_run_receipt: DryRunReceiptSummary | None
    execution: ExecutionSummary | None


class PublicationMutationResponse(PublicationSchema):
    publication: PublicationSummary | None = None
    operation: PublicationOperation | None = None


class PublicationSummaryRead(PublicationSchema):
    publication: PublicationSummary


class PublicationOperationRead(PublicationSchema):
    operation: PublicationOperation
    publication: PublicationSummary


class PublicationEntityFinding(PublicationSchema):
    finding_id: str
    code: str
    severity: Literal["error", "warning", "info"]
    message: str
    gate: bool


class PublicationEntity(PublicationSchema):
    entity_id: str
    entity_digest: str
    entity_kind: str
    label: str
    description: str | None
    target_qid: str | None
    statement_count: int
    review_status: Literal["pending", "approved", "rejected", "stale"]
    proposed_action: Literal["create", "update", "adopt", "skip", "blocked"] | None
    findings: list[PublicationEntityFinding]


class PublicationEntityPage(PublicationSchema):
    release_id: str
    release_digest: str
    items: list[PublicationEntity]
    next_cursor: str | None
    total: int


class PublicationAuditEvent(PublicationSchema):
    event_id: str
    sequence: int
    event_type: str
    occurred_at: datetime
    actor_label: str | None
    release_id: str | None
    entity_id: str | None
    message: str
    details: dict[str, object]


class PublicationAuditPage(PublicationSchema):
    publication_id: str
    items: list[PublicationAuditEvent]
    next_cursor: str | None
    total: int

File: frontend/src/api/publication.ts
import {api} from "@/api/client";

export type PublicationTarget = "test" | "live";
export type PublicationProjectionSource = "legacy" | "canonical";
export type PublicationStatus =
  | "preparing"
  | "ready_for_review"
  | "reviewed"
  | "dry_run_ready"
  | "publishing"
  | "paused"
  | "completed"
  | "cancelled"
  | "failed";
export type ReviewDecision = "approve" | "reject";
export type ReviewStatus = "pending" | "approved" | "rejected" | "stale";
export type OperationStatus = "queued" | "running" | "succeeded" | "failed" | "cancelled";

export interface RunPublicationSource {
  kind: "run";
  projection_source: PublicationProjectionSource;
  approved_only: boolean;
}

export interface ReferenceOnlySelection {
  publication_id: string;
  plan_id: string;
  plan_digest: string;
  entity_keys: string[];
}

export interface PreparePublicationRequest {
  reference_only?: ReferenceOnlySelection;
  profile_id: string;
  profile_version: string;
  target: PublicationTarget;
  source: RunPublicationSource;
}

export interface PublicationFindingCounts {
  error: number;
  warning: number;
  info: number;
}

export interface ReleaseSummary {
  release_id: string;
  release_digest: string;
  revision: number;
  created_at: string;
  entity_count: number;
  finding_counts: PublicationFindingCounts;
}

export interface ApprovalSetSummary {
  approval_set_id: string;
  approval_digest: string;
  release_id: string;
  release_digest: string;
  status: ReviewStatus;
  approved_count: number;
  rejected_count: number;
  pending_count: number;
  created_at: string;
}

export interface PlanSummary {
  blocked_actions?: Array<{entity_key: string; target_qid: string | null; reason: string; consent?: PublicationForeignQidConsent | null}>;
  plan_id: string;
  plan_digest: string;
  release_id: string;
  release_digest: string;
  approval_set_id: string;
  status: "ready" | "blocked" | "expired" | "executed";
  expires_at: string | null;
  action_counts: Record<string, number>;
}

export interface DryRunReceiptSummary {
  dry_run_receipt_id: string;
  receipt_digest: string;
  plan_id: string;
  plan_digest: string;
  status: "valid" | "stale" | "failed";
  checked_at: string;
  expires_at: string | null;
}

export interface ExecutionSummary {
  execution_id: string;
  plan_id: string;
  status: OperationStatus | "paused";
  processed: number;
  total: number;
  succeeded: number;
  failed: number;
  skipped: number;
  current_entity_label: string | null;
  started_at: string | null;
  finished_at: string | null;
}

export interface PublicationSummary {
  publication_id: string;
  run_id: string;
  profile_id: string;
  profile_version: string;
  target: PublicationTarget;
  status: PublicationStatus;
  source_current: boolean;
  current_release: ReleaseSummary;
  approval_set: ApprovalSetSummary | null;
  plan: PlanSummary | null;
  dry_run_receipt: DryRunReceiptSummary | null;
  execution: ExecutionSummary | null;
}

export interface PublicationEntityFinding {
  finding_id: string;
  code: string;
  severity: "error" | "warning" | "info";
  message: string;
  gate: boolean;
}

export interface PublicationEntity {
  reference_only?: boolean;
  deferred?: boolean;
  policy_reason?: string;
  deferred_statements?: Array<Record<string, unknown>>;
  entity_id: string;
  entity_digest: string;
  entity_kind: string;
  label: string;
  description: string | null;
  target_qid: string | null;
  statement_count: number;
  review_status: ReviewStatus;
  proposed_action: "create" | "update" | "adopt" | "skip" | "blocked" | null;
  findings: PublicationEntityFinding[];
}

export type PublicationEntitySelection =
  | {mode: "eligible_release"}
  | {mode: "entities"; entity_keys: string[]};

export interface ReviewCommand {
  type: "review";
  release_id: string;
  expected_release_digest: string;
  selection: PublicationEntitySelection;
  decision: ReviewDecision;
  reason: string;
}

export interface PublicationForeignQidConsent {
  entity_key: string;
  qid: string;
  remote_revision: number;
  entity_digest: string;
}

export interface DryRunCommand {
  force_refresh?: boolean;
  foreign_qid_consents?: PublicationForeignQidConsent[];
  type: "dry_run";
  approval_set_id: string;
  expected_approval_digest: string;
}

export interface PublishCommand {
  type: "publish";
  plan_id: string;
  dry_run_receipt_id: string;
  expected_receipt_digest: string;
}

export interface ResumeCommand {
  type: "resume";
  execution_id: string;
}

export interface CancelCommand {
  type: "cancel";
  operation_id?: string;
  reason?: string;
}

export type PublicationAdvanceCommand =
  | ReviewCommand
  | DryRunCommand
  | PublishCommand
  | ResumeCommand
  | CancelCommand;

export interface PublicationOperation {
  operation_id: string;
  command: "prepare" | PublicationAdvanceCommand["type"];
  status: OperationStatus;
  progress: ExecutionSummary | null;
  error: string | null;
}

export interface PreparePublicationResponse {
  publication: PublicationSummary | null;
  operation?: PublicationOperation | null;
}

export interface AdvancePublicationResponse {
  publication: PublicationSummary;
  operation?: PublicationOperation | null;
}

export interface PublicationSummaryQuery {
  type: "summary";
}

export interface PublicationEntitiesQuery {
  type: "entities";
  release_id: string;
  cursor?: string | null;
  limit: number;
  entity_kind?: string;
  review_status?: ReviewStatus;
  query?: string;
}

export interface PublicationOperationQuery {
  type: "operation";
  operation_id: string;
}

export interface PublicationAuditQuery {
  type: "audit";
  cursor?: string | null;
  limit: number;
}

export type PublicationReadQuery =
  | PublicationSummaryQuery
  | PublicationEntitiesQuery
  | PublicationOperationQuery
  | PublicationAuditQuery;

export interface PublicationSummaryRead {
  publication: PublicationSummary;
}

export interface PublicationEntityPage {
  release_id: string;
  release_digest: string;
  items: PublicationEntity[];
  next_cursor: string | null;
  total: number;
}

export interface PublicationOperationRead {
  operation: PublicationOperation;
  publication: PublicationSummary;
}

export interface PublicationAuditEvent {
  event_id: string;
  sequence: number;
  event_type: string;
  occurred_at: string;
  actor_label: string | null;
  release_id: string | null;
  entity_id: string | null;
  message: string;
  details: Record<string, unknown>;
}

export interface PublicationAuditPage {
  publication_id: string;
  items: PublicationAuditEvent[];
  next_cursor: string | null;
  total: number;
}

export type PublicationReadResponse =
  | PublicationSummaryRead
  | PublicationEntityPage
  | PublicationOperationRead
  | PublicationAuditPage;

function publicationPath(runId: string, publicationId?: string): string {
  const run = encodeURIComponent(runId);
  if (!publicationId) return `/runs/${run}/wikidata-publications/prepare`;
  return `/runs/${run}/wikidata-publications/${encodeURIComponent(publicationId)}`;
}

function readPublication(
  runId: string,
  publicationId: string,
  query: PublicationSummaryQuery,
): Promise<PublicationSummaryRead>;
function readPublication(
  runId: string,
  publicationId: string,
  query: PublicationEntitiesQuery,
): Promise<PublicationEntityPage>;
function readPublication(
  runId: string,
  publicationId: string,
  query: PublicationOperationQuery,
): Promise<PublicationOperationRead>;
function readPublication(
  runId: string,
  publicationId: string,
  query: PublicationAuditQuery,
): Promise<PublicationAuditPage>;
function readPublication(
  runId: string,
  publicationId: string,
  query: PublicationReadQuery,
): Promise<PublicationReadResponse> {
  return api.post<PublicationReadResponse>(
    `${publicationPath(runId, publicationId)}/read`,
    {query},
  );
}

export const PublicationApi = {
  latest: (runId: string) => api.get<PreparePublicationResponse>(`/runs/${encodeURIComponent(runId)}/wikidata-publications/latest`),
  prepare: (runId: string, request: PreparePublicationRequest) =>
    api.post<PreparePublicationResponse>(publicationPath(runId), request),

  advance: (
    runId: string,
    publicationId: string,
    command: PublicationAdvanceCommand,
  ) => api.post<AdvancePublicationResponse>(
    `${publicationPath(runId, publicationId)}/advance`,
    {command},
  ),

  read: readPublication,
};

export interface PublicationAiReviewState {
  phase?: string | null;
  message?: string | null;
  job_id: string | null;
  status: "queued" | "running" | "succeeded" | "failed" | "cancelled" | null;
  processed: number;
  total: number;
  error: string | null;
  report: {
    publication_id: string;
    plan_id: string;
    plan_digest: string;
    release_digest: string;
    tier_model: string;
    created_at: string;
    automatic?: boolean;
    policy_version?: string;
    result_publication_id?: string | null;
    items: Array<{
      entity_key: string;
      label: string;
      qid: string | null;
      status: "recommended" | "review_required" | "lookup_resolved" | "error" | "reuse_existing" | "create" | "deferred";
      reason: string;
      resolution?: {retryable?: boolean};
      consent: PublicationForeignQidConsent | null;
    }>;
  } | null;
}

export const PublicationAiReviewApi = {
  read: (runId: string, publicationId: string) =>
    api.get<PublicationAiReviewState>(`${publicationPath(runId, publicationId)}/ai-review`),
  start: (runId: string, publicationId: string, request: {
    plan_id: string; plan_digest: string; tier_model: string; force_refresh: boolean;
    automatic?: boolean; verification_model?: string;
  }) => api.post<PublicationAiReviewState>(`${publicationPath(runId, publicationId)}/ai-review`, request),
};
